src/compiler/bundle/app-data-plugin.ts

```diff
@@ -66,31 +68,53 @@ export const appDataPlugin = (
         appendEnv(config, s);
         return s.toString();
       }
-      return null;
+      if (id !== config.globalScript) {
+        return null;
+      }
+
+      const module = compilerCtx.moduleMap.get(config.globalScript);
+      if (!module.sourceMapFileText) return { code: module.staticSourceFileText, map: null };
+
+      const sourceMap: d.SourceMap = JSON.parse(module.sourceMapFileText);
+      sourceMap.sources = sourceMap.sources.map((src) => basename(src));
+      return { code: module.staticSourceFileText, map: sourceMap };
     },
 
-    transform(code, id) {
+    transform(code: string, id: string): TransformResult {
       id = normalizePath(id);
       if (globalScripts.some((s) => s.path === id)) {
         const program = this.parse(code, {});
         const needsDefault = !(program as any).body.some((s: any) => s.type === 'ExportDefaultDeclaration');
         const defaultExport = needsDefault ? '\nexport const globalFn = () => {};\nexport default globalFn;' : '';
-        code = getContextImport(platform) + code + defaultExport;
+
+        var codeMs = new MagicString(code);
+        codeMs.prepend(getContextImport(platform));
+        codeMs.append(defaultExport);
 
         const compilerOptions: ts.CompilerOptions = { ...config.tsCompilerOptions };
         compilerOptions.module = ts.ModuleKind.ESNext;
 
-        const results = ts.transpileModule(code, {
+        const results = ts.transpileModule(codeMs.toString(), {
           compilerOptions,
           fileName: id,
           transformers: {
             after: [removeCollectionImports(compilerCtx)],
           },
         });
-
+        const sourceMap = results.sourceMapText ? JSON.parse(results.sourceMapText) : null;
         buildCtx.diagnostics.push(...loadTypeScriptDiagnostics(results.diagnostics));
 
-        return results.outputText;
+        if (config.sourceMap) {
+          const codeMap = codeMs.generateMap({
+            source: id,
+            file: id + '.map',
+            includeContent: true,
+            hires: true,
+          });
+          return { code: results.outputText, map: sourceMapMerge(codeMap, sourceMap) };
```

Comment:
@johnjenkins Do you recall why we want to merge the sourcemaps here?

I did some digging, at comparing the sourcemap generated by the TS compiler (left in the image below) and the one generated by MagicString (right in the image below):

![Screen Shot 2021-09-21 at 2 05 02 PM](https://user-images.githubusercontent.com/1930213/134224207-7126ecc6-d835-4367-92da-d9759133fa66.png)

It seems the only difference is the name of `file` and `mappings`.

I _think_ per the [spec for sourcemaps](https://sourcemaps.info/spec.html#h.qz3o9nc69um5), `file` should point to `globals.js` in this situation (and should be easy enough to handle).

The `hires` (high resolution) mappings are what I'm intrigued by. I took at look at the [MagicString Docs](https://github.com/Rich-Harris/magic-string#sgeneratemap-options-)

> hether the mapping should be high-resolution. Hi-res mappings map every single character, meaning (for example) your devtools will always be able to pinpoint the exact location of function calls and so on. With lo-res mappings, devtools may only be able to identify the correct line - but they're quicker to generate and less bulky. If sourcemap 

Was the idea here to get sourcemaps with higher fidelity, then merge them back into the ones generated by TS?

Reply:
Hi @rwaskiewicz - yes the difference in mappings is the important thing ... the `file` is incorrect - or not as important at least.
We need to merge sourcemaps because we need to account for the extra import / export strings of code that can get added and to this end I needed to get creative!
1) Turn the incoming TS string to a MagicString - this will track the changes we make to the original TS and can generate a sourcemap tracking these changes.
2) Compile the TS using the `transpileModule` - this generates another sourcemap however the typescript compiler has no idea of the earlier amends made to the source TS therefore the sourcemap generated does not reflect these amends and is incorrect.
3) Merge the original sourcemap from MagicString and the new one from typescript - all changes made to the original source now being accounted for.

This is a little convoluted I agree - I was just working with what was there originally and appending strings of code to source input should be avoided in general for this reason.
For future devs - try to make these amends to sourcecode within TS plugins / AST to avoid this hinkiness.

Regarding the `hires` property - I simply applied it because the weird process described and corresponding merge would produce slightly incorrect sourcemaps and this fixed it. 
Attached is 2 screen caps. 1 with `hires` - each console points to the correct line from the original source. 2 without `hires` source lines highlighted can be out by 1
https://user-images.githubusercontent.com/5030133/134250837-edc67824-f668-4b85-8b34-4966017cc70c.mov
https://user-images.githubusercontent.com/5030133/134250827-634f7e57-5e24-451c-a353-21dde518489b.mov